### PR TITLE
Update Typescript

### LINKED ISSUE #2161: Include type definitions in Pyodide npm package
## 🚀 Feature

Pyodide's JavaScript API is implemented in TypeScript, but the [npm package](https://www.npmjs.com/package/pyodide) doesn't contain any type definitions. How about including type definitions in the npm package? 

### Motivation

If a package doesn't have type definitions, it is hard to use.

### Pitch

I want to import type definitions from Pyodide npm package.

### Alternatives

Write type definitions in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)?

### Additional context

**Current (0.19.0) `pyodide` in `node_modules`**
image
[`index.d.ts`](https://github.com/pyodide/pyodide/blob/main/src/js/index.d.ts) is almost empty.

### PR DESCRIPTION
**Pull request by jasongrout**
### Description

This updates Typescript to the latest version and adds a few stricter options to check code.

It also enables generating the tsd files, helping fix #2161. 

This builds on top of #2471

### CODE REVIEW

File: src/js/package.json [hunk 1]
```diff
@@ -105,7 +105,7 @@
   "tsd": {
     "compilerOptions": {
       "lib": [
-        "ES2018",
+        "ES2017",
```
Comment (hoodmane) [inline src/js/package.json:108]: What is the point of this? Are you just setting this to the minimum value that allows the build to succeed?

**Review by hoodmane — APPROVED** [review]
Thanks again @jasongrout! This is super helpful.

File: src/js/package.json (hunk 1 above)
Comment (jasongrout) [reply src/js/package.json:108]: Just making things consistent between typescript and tsd configs.

File: src/core/pyproxy.ts [hunk 2]
```diff
@@ -805,6 +805,9 @@ export class PyProxyIteratorMethods {
       Hiwire.decref(idarg);
     }
     let HEAPU32 = Module.HEAPU32;
+    // HEAPU32 is used in the DEREF_U32 C preprocessor macro. Typescript doesn't know this.
+    // So we "use" HEAPU32 once to trick Typescript, so we can enable strictUnusedLocalVariables.
+    HEAPU32;
```
[the remark this answers is not in the document]
Comment (jasongrout) [reply src/core/pyproxy.ts]: Done

[the remark this answers is not in the document]
Comment (jasongrout) [reply src/core/pyproxy.ts]: When I look in the `pyproxy.gen.ts` file, I don't see HEAPU32 defined before it is used there.

[the remark this answers is not in the document]
Comment (jasongrout) [reply src/core/pyproxy.ts:810]: Actually, typescript shouldn't be processing that file anyway before the preprocessor hits it. I think it's better to handle this over in #2481, where typescript won't be processing pyproxy.ts directly. I reverted back to the previous commit we had here.

**Comment by rth** [comment]
Thanks a lot @jasongrout !

**Pull request merged**